pkg/agent/openflow/client.go

```diff
@@ -470,3 +506,108 @@ func (c *client) setupPolicyOnlyFlows() error {
+	dstMAC string,
+	srcIP string,
+	dstIP string,
+	IPProtocol uint8,
+	ttl uint8,
+	IPFlags uint16,
+	TCPsPort uint16,
+	TCPdPort uint16,
+	TCPFlags uint8,
+	UDPsPort uint16,
+	UDPdPort uint16,
+	ICMPType uint8,
+	ICMPCode uint8,
+	ICMPID uint16,
+	ICMPSequence uint16,
+	inPort uint32,
+	outPort int32) error {
+
+	regName := fmt.Sprintf("%s%d", binding.NxmFieldReg, traceflowReg)
+
+	packetOutBuilder := c.bridge.BuildPacketOut()
+	parsedSrcMAC, _ := net.ParseMAC(srcMAC)
+	parsedDstMAC, _ := net.ParseMAC(dstMAC)
+	if dstMAC == "" {
+		parsedDstMAC = globalVirtualMAC
+	}
+
+	packetOutBuilder = packetOutBuilder.SetSrcMAC(parsedSrcMAC)
+	packetOutBuilder = packetOutBuilder.SetDstMAC(parsedDstMAC)
+	packetOutBuilder = packetOutBuilder.SetSrcIP(net.ParseIP(srcIP))
+	packetOutBuilder = packetOutBuilder.SetDstIP(net.ParseIP(dstIP))
+
+	if ttl == 0 {
+		packetOutBuilder = packetOutBuilder.SetTTL(128)
+	} else {
+		packetOutBuilder = packetOutBuilder.SetTTL(ttl)
+	}
+	packetOutBuilder = packetOutBuilder.SetIPFlags(IPFlags)
+
+	switch IPProtocol {
+	case 1:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolICMP)
+		packetOutBuilder = packetOutBuilder.SetICMPType(ICMPType)
+		packetOutBuilder = packetOutBuilder.SetICMPCode(ICMPCode)
+		packetOutBuilder = packetOutBuilder.SetICMPID(ICMPID)
+		packetOutBuilder = packetOutBuilder.SetICMPSequence(ICMPSequence)
+	case 6:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolTCP)
+		if TCPsPort == 0 {
+			TCPsPort = uint16(rand.Uint32())
+		}
+		packetOutBuilder = packetOutBuilder.SetTCPSrcPort(TCPsPort)
+		packetOutBuilder = packetOutBuilder.SetTCPDstPort(TCPdPort)
+		packetOutBuilder = packetOutBuilder.SetTCPFlags(TCPFlags)
+	case 17:
+		packetOutBuilder = packetOutBuilder.SetIPProtocol(binding.ProtocolUDP)
+		packetOutBuilder = packetOutBuilder.SetUDPSrcPort(UDPsPort)
+		packetOutBuilder = packetOutBuilder.SetUDPDstPort(UDPdPort)
+	}
+
+	packetOutBuilder = packetOutBuilder.SetInport(inPort)
+	if outPort != -1 {
+		packetOutBuilder = packetOutBuilder.SetOutport(uint32(outPort))
+	}
+	packetOutBuilder = packetOutBuilder.AddLoadAction(regName, uint64(dataplaneTag), ofTraceflowMarkRange)
+
+	packetOutObj := packetOutBuilder.Done()
+	return c.bridge.SendPacketOut(packetOutObj)
+}
+
+func (c *client) InstallTraceflowFlows(dataplaneTag uint8) error {
+	flow := c.traceflowL2ForwardOutputFlow(dataplaneTag, cookie.Default)
+	if err := c.Add(flow); err != nil {
+		return err
+	}
+	flow = c.traceflowConnectionTrackFlows(dataplaneTag, cookie.Default)
+	if err := c.Add(flow); err != nil {
+		return err
+	}
+	flows := []binding.Flow{}
+	for _, ctx := range c.globalConjMatchFlowCache {
+		if ctx.dropFlow != nil {
+			flows = append(
+				flows,
+				ctx.dropFlow.CopyToBuilder(priorityNormal+2).
+					MatchRegRange(int(traceflowReg), uint32(dataplaneTag), ofTraceflowMarkRange).
+					SetHardTimeout(300).
+					Action().SendToController(1).
+					Done())
+		}
+	}
+	return c.AddAll(flows)
```

Comment:
Why not use a single bundle to submit all flows in this func?

Reply:
Because the 3 parts of flows are different, we can get more accurate error for each part.